server/room-battle.ts

```diff
@@ -1348,105 +1295,185 @@ export class RoomBattle extends RoomGames.RoomGame<RoomBattlePlayer> {
+			battleRoom?.sendUser(user, button);
+			return;
+		}
+
+		const cmd = `/msgroom ${room.roomid},/confirmready`;
+		const button = `|c|&|/uhtml controls,<p>Are you ready for game ${gameNum}, ${this.name}?</p>` +
+			(this.ready ?
+				`<button class="button" disabled><i class="fa fa-check"></i> I'm ready!</button> &ndash; waiting for opponent...` :
+				`<button class="button notifying" name="send" value="${cmd}">I'm ready!</button>`);
+		this.sendRoom(button);
+		battleRoom?.sendUser(user, button);
+	}
+}
+
+export class BestOfGame extends RoomGame<BestOfPlayer> {
+	override allowRenames = false;
 	bestOf: number;
 	format: Format;
-	score: number[] | null = null;
 	winThreshold: number;
-	options: RoomBattleOptions;
-	p1!: ID;
-	p2!: ID;
-	wins = {p1: 0, p2: 0};
-	ready: {p1: boolean, p2: boolean} | null = null;
+	options: Omit<RoomBattleOptions, 'players'> & {players: null};
 	ties = 0;
-	games: {battle: Room, winner: string | null, rated: number}[] = [];
-	allowRenames = false;
+	games: {room: Room, winner: BestOfPlayer | null | undefined, rated: number}[] = [];
 	playerNum = 0;
-	winner: ID | null = null;
-	waitingBattle: RoomBattle | null = null;
-	nextBattleTimerStart: number | null = null;
+	/** null = tie, undefined = not ended */
+	winner: BestOfPlayer | null | undefined = undefined;
+	/** when waiting between battles, this is the just-ended battle room, the one with the |tempnotify| */
+	waitingBattle: GameRoom | null = null;
+	nextBattleTimerEnd: number | null = null;
 	nextBattleTimer: NodeJS.Timer | null = null;
-	ended = false;
 	needsTimer = false;
-	teams: Record<string, PokemonSet[] | null | undefined> = {};
 	constructor(room: Room, options: RoomBattleOptions) {
-		super(room, false);
+		super(room);
 		this.format = Dex.formats.get(options.format);
 		this.bestOf = Number(Dex.formats.getRuleTable(this.format).valueRules.get('bestof'))!;
 		this.winThreshold = Math.floor(this.bestOf / 2) + 1;
 		this.title = this.format.name;
 		if (!toID(this.title).includes('bestof')) {
 			this.title += ` (Best-of-${this.bestOf})`;
 		}
-		this.options = options;
-		options.isSubBattle = true;
-		if (!options.players && (options.p1 || options.p2)) {
-			options.players = [options.p1?.user, options.p2?.user].map(toID);
-		}
-		for (const userid of options.players || []) {
-			const player = this.makePlayer(userid);
-			this.playerTable[userid] = player;
-			this.room.auth.set(userid, Users.PLAYER_SYMBOL);
-			this[`p${player.num}` as 'p1' | 'p2'] = userid;
+		this.options = {
+			...options,
+			isSubBattle: true,
+			allowRenames: false,
+			players: null,
+		};
+		for (const playerOpts of options.players) {
+			this.addPlayer(playerOpts.user, playerOpts);
 		}
 		process.nextTick(() => this.nextGame());
 	}
-	onConnect(user: User) {
+	override onConnect(user: User) {
 		const player = this.playerTable[user.id];
-		if (player) {
-			player.sendRoom('|cantleave|');
-		}
+		player?.sendRoom('|cantleave|');
+		player?.readyButton();
```

Comment:
@DaWoblefet You mentioned a "bug" about how we didn't support sending private information on join, but we do. There are two ways to do it, one is to send it directly in `onConnect`, and one is the system we use in battles with `|split`.

Reply:
``|split`` doesn't work because if you refresh, you initially join the room as a Guest account, then it switches to the player name. It actually does work if the room is private though, because it forces you to authenticate with a username before it lets you in. Sample behavior with exact damage rolls, which uses ``|split``. That's what I meant by it being "bugged".

https://github.com/smogon/pokemon-showdown/assets/23667022/68b4bd06-942b-46c4-8a1c-9a9b88d144f2

Reply:
Ah, yeah, you're talking about #8954. It's also easy enough to fix, something along the lines of making all battles work like private rooms. Or perhaps just force a rejoin when you log in. Anyway, doable, and also not relevant to BestOfGame.

Reply:
Although when I say "easy enough" _here_, it's not a "I can do it instantly" kind of fix like with the timer. It's still gonna have to go way on the back burner.